ENH: Add CAT12 interfaces

mfmachado opened the pull request:
<!--

Pull-request guidelines
-----------------------

1. If you would like to list yourself as a Nipype contributor and your name is not mentioned please modify .zenodo.json file.
2. By submitting this request you acknowledge that your contributions are available under the Apache 2 license.
3. Use a descriptive prefix for your PR: ENH (enhancement), FIX, TST, DOC, STY, REF (refactor), WIP (Work in progress)
4. Run `make check-before-commit` before submitting the PR.

-->
## Summary
<!-- Please reference any related issue and use fixes/close to automatically
close them, if pertinent -->

Fixes #2783 .

## List of changes proposed in this PR (pull-request)
<!-- We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

## Acknowledgment

- [x] \(Mandatory\) I acknowledge that this contribution will be available under the Apache 2 license.

effigies commented on nipype/interfaces/cat12/preprocess.py:
```diff
@@ -0,0 +1,363 @@
+import os
+import traits
+from nipype.interfaces.base import InputMultiPath, TraitedSpec, isdefined
```
We use `nipype.interfaces.base` to override some aspects of traits. I would suggest the following:

```suggestion
from nipype.interfaces.base import InputMultiPath, TraitedSpec, isdefined, traits, File, Str
```

mfmachado replied:
Hi @effigies sorry for the late reply, I only had the time to correct you reviews today. I do not known what you mean by make specs...

effigies commented on nipype/interfaces/cat12/preprocess.py:
```diff
@@ -0,0 +1,363 @@
+from nipype.interfaces.spm import SPMCommand
+from nipype.interfaces.spm.base import SPMCommandInputSpec, ImageFileSPM, scans_for_fnames, scans_for_fname
+from nipype.utils.filemanip import split_filename, fname_presuffix
+from traits.trait_types import Int, File
+from traits.trait_types import List
```
```suggestion
```

effigies commented on nipype/interfaces/cat12/preprocess.py:
```diff
@@ -0,0 +1,363 @@
+
+
+class CAT12SegmentInputSpec(SPMCommandInputSpec):
+    """
+    CAT12: Segmentation
+    This toolbox is an extension to the default segmentation in SPM12, but uses a completely different segmentation
+    approach.
+    The segmentation approach is based on an Adaptive Maximum A Posterior (MAP) technique without the need for a priori
+    information about tissue probabilities. That is, the Tissue Probability Maps (TPM) are not used constantly in the
+    sense of the classical Unified Segmentation approach (Ashburner et. al. 2005), but just for spatial normalization.
+    The following AMAP estimation is adaptive in the sense that local variations of the parameters (i.e., means and
+    variance) are modeled as slowly varying spatial functions (Rajapakse et al. 1997). This not only accounts for
+    intensity inhomogeneities but also for other local variations of intensity.
+    Additionally, the segmentation approach uses a Partial Volume Estimation (PVE) with a simplified mixed model of at
+    most two tissue types (Tohka et al. 2004). We start with an initial segmentation into three pure classes: gray
+    matter (GM), white matter (WM), and cerebrospinal fluid (CSF) based on the above described AMAP estimation. The
+    initial segmentation is followed by a PVE of two additional mixed classes: GM-WM and GM-CSF. This results in an
+    estimation of the amount (or fraction) of each pure tissue type present in every voxel (as single voxels - given by
+    Another important extension to the SPM12 segmentation is the integration of the Dartel or Geodesic Shooting
+    registration into the toolbox by an already existing Dartel/Shooting template in MNI space. This template was
+    derived from 555 healthy control subjects of the IXI-database (http://www.brain-development.org) and provides the
+    several Dartel or Shooting iterations. Thus, for the majority of studies the creation of sample-specific templates
+    is not necessary anymore and is mainly recommended for children data.'};
+
+    http://www.neuro.uni-jena.de/cat12/CAT12-Manual.pdf#page=15
+
+    Examples
+    --------
+    >>> path_mr = 'structural.nii'
+    >>> cat = CAT12Segment(in_files=path_mr)
+    >>> cat.run() # doctest: +SKIP
+    """
```
This docstring should go on `CAT12Segment`, not `CAT12SegmentInputSpec`.

effigies commented on nipype/interfaces/cat12/preprocess.py:
```diff
@@ -0,0 +1,363 @@
+    in_files = InputMultiPath(ImageFileSPM(exists=True), field="data", desc="file to segment", mandatory=True,
+                              copyfile=False)
+
+    help_tpm = 'Tissue Probability Map. Select the tissue probability image that includes 6 tissue probability ' \
+               'classes for (1) grey matter, (2) white matter, (3) cerebrospinal fluid, (4) bone, (5) non-brain soft ' \
+               'tissue, and (6) the background.  CAT uses the TPM only for the initial SPM segmentation.'
+    tpm = InputMultiPath(ImageFileSPM(exists=True), field="tpm", desc=help_tpm, mandatory=False,
+                         copyfile=False)
```
For variables that are not traits, we use `_` as a prefix to mark them as private. This will help declutter tab-completion in IPython or similar environments.

```suggestion
    _help_tpm = 'Tissue Probability Map. Select the tissue probability image that includes 6 tissue probability ' \
               'classes for (1) grey matter, (2) white matter, (3) cerebrospinal fluid, (4) bone, (5) non-brain soft ' \
               'tissue, and (6) the background.  CAT uses the TPM only for the initial SPM segmentation.'
    tpm = InputMultiPath(ImageFileSPM(exists=True), field="tpm", desc=_help_tpm, mandatory=False,
                         copyfile=False)
```

And so on for other traits.

effigies commented on nipype/interfaces/cat12/preprocess.py:
```diff
@@ -0,0 +1,363 @@
+    tpm = InputMultiPath(ImageFileSPM(exists=True), field="tpm", desc=help_tpm, mandatory=False,
+                         copyfile=False)
+
+    n_jobs = traits.trait_types.Int(1, usedefault=True, mandatory=True, field="nproc", desc="Number of threads")
+    use_prior = traits.trait_types.Str(field="useprior", usedefault=True)
```
Based on changed imports:

```suggestion
    n_jobs = traits.Int(1, usedefault=True, mandatory=True, field="nproc", desc="Number of threads")
    use_prior = Str(field="useprior", usedefault=True)
```

And can you add a description to `use_prior`?

effigies commented on nipype/interfaces/cat12/preprocess.py:
```diff
@@ -0,0 +1,363 @@
+                      'Values:\nOptimal: [1.0 0.1]\nFixed 1.0 mm: [1.0 0.1];\nFixed 0.8 mm:[0.8 0.1]' \
+                      '\nBest native: [0.5 0.1]'
+    internal_resampling_process = traits.trait_types.Tuple(traits.trait_types.Float(1), traits.trait_types.Float(0.1),
+                                                           minlen=2, maxlen=2,
```
I'm pretty sure this is redundant, but you may want to check before applying this.

```suggestion
```

effigies commented on nipype/interfaces/cat12/preprocess.py:
```diff
@@ -0,0 +1,363 @@
+        f = self.inputs.in_files[0]
+        pth, base, ext = split_filename(f)
+
+        outputs["mri_images"] = [os.path.join(os.path.join(pth, "mri"), f) for f in
+                                 os.listdir(os.path.join(pth, "mri"))
+                                 if os.path.isfile(os.path.join(os.path.join(pth, "mri"), f))]
```
`os.path.join()` takes any number of arguments.

```suggestion
        outputs["mri_images"] = [os.path.join(pth, "mri", f) for f in
                                 os.listdir(os.path.join(pth, "mri"))
                                 if os.path.isfile(os.path.join(pth, "mri", f))]
```

But you can actually do this more succinctly with the `pathlib` library:

```suggestion
        outputs["mri_images"] = [str(mri) for mri in Path(pth).glob("mri/*") if mri.is_file()]
```

For the remaining examples, I'll use that idiom.

effigies commented on nipype/interfaces/cat12/preprocess.py:
```diff
@@ -0,0 +1,363 @@
+
+            for idx, (image, prefix) in enumerate([("modulated", "mw"), ("dartel", "r"), ("native", "")]):
+                outtype = f'{tissue}_output_{image}'
+                if isdefined(getattr(self.inputs, outtype)) and getattr(self.inputs, outtype):
```
`bool(Undefined) == False`, so this can just be:

```suggestion
                if getattr(self.inputs, outtype):
```

effigies commented on nipype/interfaces/cat12/preprocess.py:
```diff
@@ -0,0 +1,363 @@
+
+        for tidx, tissue in enumerate(["gm", "wm", "csf"]):
+
+            for idx, (image, prefix) in enumerate([("modulated", "mw"), ("dartel", "r"), ("native", "")]):
```
Unused `idx`:

```suggestion
            for image, prefix in (("modulated", "mw"), ("dartel", "r"), ("native", "")):
```

effigies commented on nipype/interfaces/cat12/preprocess.py:
```diff
@@ -0,0 +1,363 @@
+        if isdefined(self.inputs.save_bias_corrected) and self.inputs.save_bias_corrected:
+            outputs["bias_corrected_image"] = fname_presuffix(f, prefix=os.path.join("mri", 'wmi'))
+
+        outputs["surface_files"] = [os.path.join(os.path.join(pth, "surf"), f) for f in
+                                    os.listdir(os.path.join(pth, "surf"))
+                                    if os.path.isfile(os.path.join(os.path.join(pth, "surf"), f))]
```
```suggestion
        outputs["surface_files"] = [str(surf) for surf in Path(pth).glob("surf/*") if surf.is_file()]
```

effigies commented on nipype/interfaces/cat12/preprocess.py:
```diff
@@ -0,0 +1,363 @@
+                                    os.listdir(os.path.join(pth, "surf"))
+                                    if os.path.isfile(os.path.join(os.path.join(pth, "surf"), f))]
+
+        for tidx, hemisphere in enumerate(["rh", "lh"]):
+            for idx, suffix in enumerate(["central", "sphere"]):
```
Unused indices.

```suggestion
        for hemisphere in ("rh", "lh"):
            for suffix in ("central", "sphere"):
```

effigies reviewed (commented):
I had a look through preprocess.py, but haven't done the other yet. Don't forget to `make specs`.

effigies commented on nipype/interfaces/cat12/preprocess.py:
```diff
@@ -0,0 +1,355 @@
+                         copyfile=False)
+
+    n_jobs = traits.Int(1, usedefault=True, mandatory=True, field="nproc", desc="Number of threads")
+    use_prior = Str(field="useprior", usedefault=True)
```
Missing descriptions for:
* use_prior
* surface_measures
* neuromorphometrics
* lpba40
* cobra
* hammers
* save_bias_corrected

Unclear descriptions:
* own_atlas
* gm_output_*
* wm_output_*
* csf_output_*
* label_* / output_labelnative
* las_*

effigies commented:
There's a Makefile with a `specs` entry:

https://github.com/nipy/nipype/blob/4e8f54d8e304dd940cfdc5d42f37288651bc8a03/Makefile#L71-L73

If you run `make specs`, it will auto-generate a test for your new interface. If you don't have `make` installed, you can just run `python tools/checkspecs.py`, as it does the same thing.

effigies commented:
@mfmachado I went ahead and merged `master`, ran `black` and `make specs` on your PR to get it ready to finish. Are you good with it at this point? If so, I can aim to do a final review soon.

effigies commented on nipype/interfaces/cat12/surface.py:
```diff
@@ -0,0 +1,294 @@
+import os
+from pathlib import Path
+
+from traits.trait_base import _Undefined
+
+from nipype.interfaces.base import File, InputMultiPath, TraitedSpec, traits
```
We have the `isdefined` function.

```suggestion
from nipype.interfaces.base import File, InputMultiPath, TraitedSpec, traits, isdefined
```

effigies commented on nipype/interfaces/cat12/surface.py:
```diff
@@ -0,0 +1,294 @@
+                        name_hemisphere = hemisphere + "_" + parameter_name
+                        if isinstance(outputs[name_hemisphere], _Undefined):
+                            outputs[name_hemisphere] = []
+                        if isinstance(outputs[all_files_hemisphere], _Undefined):
```
```suggestion
                        if not isdefined(outputs[all_files_hemisphere]):
```

effigies reviewed (commented):
This is basically good to go. Some small comments, only.

mfmachado commented:
Hi @effigies,

I have been super busy in the last couple of weeks, when are you going to do the release? I can try to do it this weekend.

effigies commented:
Happy to wait on this PR if you expect to get it in soon. If you're very busy, we can also release and wait on this until you have more time.

mfmachado commented:
Hi @effigies thanks for you input :) 
I updated the code and corrected a bug that I detected meanwhile regarding the TPM and Shooting template. I use this code regularly, so I hope it does not have bugs.

mfmachado commented:
Hi @effigies all testes passed :D

effigies reviewed (approved):
Thanks! I did a few minor cleanups just to polish this off, rather than put you through another round of review. If you want to have a quick look, let me know if there's anything you disagree with. Otherwise, I'll plan to merge tomorrow.

mfmachado commented:
hi, yes I agree with your changes :)
